dali/python/nvidia/dali/pipeline.py

```diff
@@ -293,16 +326,28 @@ def _prepare_graph(self):
 
         # Add the ops to the graph and build the backend
         related_logical_id = {}
+        self._ops = []
         while ops:
             op = ops.pop()
+            self._ops.append(op)
             if op.relation_id not in related_logical_id:
                 related_logical_id[op.relation_id] = self._pipe.AddOperator(op.spec, op.name)
             else:
                 self._pipe.AddOperator(op.spec, op.name, related_logical_id[op.relation_id])
         self._prepared = True
+        self._setup_input_callbacks()
         self._names_and_devices = [(e.name, e.device) for e in outputs]
 
-    def build(self):
+    def _setup_input_callbacks(self):
+        groups = set()
+        for op in self._ops:
+            callback = getattr(op._op, "_callback", None)
```

Comment:
Can we do this in a clearer way, or maybe explain that this looks for ExternalSource Ops that have the callback set?
Now it requires the reader to know that those ops have this member.
Shouldn't we filter the Ops for ExternalSource directly or something?

Reply:
I'll add the comment for the loop:
```
# Find groups of ExternalSources which should be populated by a single callback
```
Also, maybe I'll find a better way to determine whether given op is an external source.